src/core/lib/event_engine/posix_engine/posix_endpoint.cc

```diff
@@ -0,0 +1,1135 @@
+    GPR_ASSERT(poller_->CanTrackErrors());
+  }
+
+  bool flush_result = zerocopy_send_record != nullptr
+                          ? TcpFlushZerocopy(zerocopy_send_record, status)
+                          : TcpFlush(status);
+  if (!flush_result) {
+    Ref();
+    write_cb_ = std::move(on_writable);
+    current_zerocopy_send_ = zerocopy_send_record;
+    handle_->NotifyOnWrite(on_write_);
+  } else {
+    on_writable(status);
+  }
+}
+
+void PosixEndpointImpl::MaybeShutdown(absl::Status why) {
+  if (poller_->CanTrackErrors()) {
+    ZerocopyDisableAndWaitForRemaining();
+    stop_error_notification_.store(true, std::memory_order_release);
+    handle_->SetHasError();
+  }
+  handle_->ShutdownHandle(why);
+  Unref();
+}
+
+PosixEndpointImpl ::~PosixEndpointImpl() {
+  handle_->OrphanHandle(on_done_, nullptr, "");
+  delete on_read_;
+  delete on_write_;
+  delete on_error_;
+}
+
+PosixEndpointImpl::PosixEndpointImpl(EventHandle* handle,
+                                     PosixEngineClosure* on_done,
+                                     std::shared_ptr<EventEngine> engine,
+                                     const PosixTcpOptions& options)
+    : sock_(PosixSocketWrapper(handle->WrappedFd())),
+      on_done_(on_done),
+      traced_buffers_(),
+      handle_(handle),
+      poller_(handle->Poller()),
+      engine_(engine) {
+  PosixSocketWrapper sock(handle->WrappedFd());
+  fd_ = handle_->WrappedFd();
+  GPR_ASSERT(options.resource_quota != nullptr);
+  memory_owner_ = options.resource_quota->memory_quota()->CreateMemoryOwner(
+      *sock.PeerAddressString());
+  self_reservation_ = memory_owner_.MakeReservation(sizeof(PosixEndpointImpl));
+  local_address_ = *sock.LocalAddress();
+  peer_address_ = *sock.PeerAddress();
+  target_length_ = static_cast<double>(options.tcp_read_chunk_size);
+  bytes_read_this_round_ = 0;
+  min_read_chunk_size_ = options.tcp_min_read_chunk_size;
+  max_read_chunk_size_ = options.tcp_max_read_chunk_size;
+  tcp_zerocopy_send_ctx_ = absl::make_unique<TcpZerocopySendCtx>(
+      options.tcp_tx_zerocopy_max_simultaneous_sends,
+      options.tcp_tx_zerocopy_send_bytes_threshold);
+  frame_size_tuning_enabled_ = grpc_core::IsTcpFrameSizeTuningEnabled();
+  if (options.tcp_tx_zero_copy_enabled &&
+      !tcp_zerocopy_send_ctx_->MemoryLimited() && poller_->CanTrackErrors()) {
+#ifdef GRPC_LINUX_ERRQUEUE
+    const int enable = 1;
+    auto err =
+        setsockopt(fd_, SOL_SOCKET, SO_ZEROCOPY, &enable, sizeof(enable));
+    if (err == 0) {
+      tcp_zerocopy_send_ctx_->SetEnabled(true);
+    } else {
+      gpr_log(GPR_ERROR, "Failed to set zerocopy options on the socket.");
+    }
+#endif
+  }
+
+#ifdef GRPC_HAVE_TCP_INQ
+  int one = 1;
+  if (setsockopt(fd_, SOL_TCP, TCP_INQ, &one, sizeof(one)) == 0) {
+    inq_capable_ = true;
+  } else {
+    gpr_log(GPR_DEBUG, "cannot set inq fd=%d errno=%d", fd_, errno);
+    inq_capable_ = false;
+  }
+#else
+  inq_capable_ = false;
+#endif /* GRPC_HAVE_TCP_INQ */
```

Comment:
Please be consistent with comment styles, I'd suggest `#endif  // GRPC_HAVE_TCP_INQ` as is used elsewhere in this file.

Reply:
Fixed it now.